packages/shared/src/enums.ts

```diff
@@ -26,71 +30,130 @@ export type ChainflipNetwork =
-  [Assets.DOT]: 4,
-  [Assets.BTC]: 5,
-};
-
-export const chainAssets = {
-  [Chains.Ethereum]: [Assets.ETH, Assets.USDC, Assets.FLIP],
-  [Chains.Bitcoin]: [Assets.BTC],
-  [Chains.Polkadot]: [Assets.DOT],
-} satisfies Record<Chain, Asset[]>;
-
-export const chainNativeAssets = {
-  [Chains.Ethereum]: Assets.ETH,
-  [Chains.Bitcoin]: Assets.BTC,
-  [Chains.Polkadot]: Assets.DOT,
-} satisfies Record<Chain, Asset>;
+export const assetConstants = {
+  [InternalAssets.Eth]: {
+    chain: Chains.Ethereum,
+    asset: Assets.ETH,
+    name: 'Ether',
+    decimals: 18,
+    contractId: 1,
+  },
+  [InternalAssets.Flip]: {
+    chain: Chains.Ethereum,
+    asset: Assets.FLIP,
+    name: 'FLIP',
+    decimals: 18,
+    contractId: 2,
+  },
+  [InternalAssets.Usdc]: {
+    chain: Chains.Ethereum,
+    asset: Assets.USDC,
+    name: 'USDC',
+    decimals: 6,
+    contractId: 3,
+  },
+  [InternalAssets.Dot]: {
+    chain: Chains.Polkadot,
+    asset: Assets.DOT,
+    name: 'Polkadot',
+    decimals: 10,
+    contractId: 4,
+  },
+  [InternalAssets.Btc]: {
+    chain: Chains.Bitcoin,
+    asset: Assets.BTC,
+    name: 'Bitcoin',
+    decimals: 8,
+    contractId: 5,
+  },
+} satisfies Record<
+  InternalAsset,
+  {
+    chain: Chain;
+    asset: Asset;
+    name: string;
+    decimals: number;
+    contractId: number;
+  }
+>;
 
 // https://github.com/chainflip-io/chainflip-backend/blob/a2a3c2e447e7b629c4b96797d9eed22eb5b87a0b/state-chain/primitives/src/chains.rs#L52-L56
-export const chainContractIds: Record<Chain, number> = {
-  [Chains.Ethereum]: 1,
-  [Chains.Polkadot]: 2,
-  [Chains.Bitcoin]: 3,
-};
-
+export const chainConstants = {
+  [Chains.Ethereum]: {
+    assets: [Assets.ETH, Assets.FLIP, Assets.USDC],
+    gasAsset: Assets.ETH,
+    contractId: 1,
+  },
+  [Chains.Polkadot]: {
+    assets: [Assets.DOT],
+    gasAsset: Assets.DOT,
+    contractId: 2,
+  },
+  [Chains.Bitcoin]: {
+    assets: [Assets.BTC],
+    gasAsset: Assets.BTC,
+    contractId: 3,
+  },
+} satisfies Record<
+  Chain,
+  { assets: Asset[]; gasAsset: Asset; contractId: number }
+>;
+
+export type AssetOfChain<C extends Chain> =
+  (typeof chainConstants)[C]['assets'][number];
+
+export type UncheckedAssetAndChain = { asset: Asset; chain: Chain };
 export type AssetAndChain = {
-  [A in Asset]: { asset: A; chain: (typeof assetChains)[A] };
-}[Asset];
+  [C in Chain]: { chain: C; asset: AssetOfChain<C> };
+}[Chain];
 
-export type UncheckedAssetAndChain = {
-  asset: Asset;
-  chain: Chain;
+export type ChainAssetMap<T> = {
+  [C in Chain]: {
+    [A in AssetOfChain<C>]: T;
+  };
 };
 
-function assertAsset(asset: string): asserts asset is Asset {
-  if (!(asset in Assets)) throw new Error('invalid asset');
+export function isValidAssetAndChain(
+  assetAndChain: UncheckedAssetAndChain,
+): assetAndChain is AssetAndChain {
+  const { asset, chain } = assetAndChain;
+  const validAssets = chainConstants[chain].assets as string[];
+
+  return validAssets.includes(asset);
```

Comment:
This isn't a safe cast because we haven't validated that `chain` is actually a chain
```suggestion
  if (!isValidChain(chain)) return false;
  const validAssets = chainConstants[chain].assets;

  return validAssets.includes(asset);
```

Reply:
not sure about this - the `chain` already has the type `Chain` here because 
```
UncheckedAssetAndChain = { asset: Asset; chain: Chain };
```
should i add an additional check?

Reply:
Yeah I guess the type is wrong then. We do use this to parse and validate user input. Maybe we should change the types there to be `string` instead, but we definitely need a check here

Reply:
changed the type to `{ asset: string; chain: string };`